File: src/mrcv-morphologyImage.cpp
#include <mrcv/mrcv.h>

namespace mrcv
{
/**
* @brief функция морфологического открытия.
* @param image - исходное фото, out - путь для нового файла, element - размер преобразования.
* @return - результат работы функции.
*/
int openingMorphological(cv::Mat image,std::string out,cv::Mat element)
{

    cv::Mat output;
    // Opening
    cv::morphologyEx(image, output,
                 cv::MORPH_OPEN, element,
                 cv::Point(-1, -1), 2);
    imwrite(out,output);
    return 0;
}
/**
* @brief функция морфологического закрытия.
* @param image - исходное фото, out - путь для нового файла, element - размер преобразования.
* @return - результат работы функции.
*/
int closingMorphological(cv::Mat image,std::string out,cv::Mat element)
{
    cv::Mat output;
    // Closing
    cv::morphologyEx(image, output,
                 cv::MORPH_CLOSE, element,
                 cv::Point(-1, -1), 2);
    imwrite(out,output);
     return 0;
}
/**
* @brief функция морфологического градиента.
* @param image - исходное фото, out - путь для нового файла, element - размер преобразования.
* @return - результат работы функции.
*/
int gradientMorphological(cv::Mat image,std::string out,cv::Mat element)
{
    cv::Mat output;
    // Gradient
    cv::morphologyEx(image, output,
                 cv::MORPH_GRADIENT, element,
                 cv::Point(-1, -1), 1);
    imwrite(out,output);
    return 0;

}
/**
* @brief функция морфологической эрозии.
* @param image - исходное фото, out - путь для нового файла, element - размер преобразования.
* @return - результат работы функции.
*/
int erodeMorphological(cv::Mat image,std::string out,cv::Mat element)
{
    cv::Mat erod;
    // For Erosion
    cv::erode(image, erod, element,
          cv::Point(-1, -1), 1);
    imwrite(out,erod);
    return 0;
}
/**
* @brief функция морфологического расширения.
* @param image - исходное фото, out - путь для нового файла, element - размер преобразования.
* @return - результат работы функции.
*/
int dilationMorphological(cv::Mat image,std::string out,cv::Mat element)
{
    cv::Mat dill;
    // For Dilation
    cv::dilate(image, dill, element,
           cv::Point(-1, -1), 1);
    imwrite(out,dill);
    return 0;
}
/**
* @brief главная функция.
* @param image - исходное фото, out - путь для нового файла, metod - метод преобразования , morph_size - размер преобразования.
* @return - результат работы функции.
*/
int morphologyImage(cv::Mat image,std::string out,mrcv::METOD_MORF metod,int morph_size)
{
    cv::Mat element = getStructuringElement(
        cv::MORPH_RECT,
        cv::Size(2 * morph_size + 1,
             2 * morph_size + 1),
        cv::Point(morph_size,
              morph_size));
    switch (metod)
    {
    case mrcv::METOD_MORF::OPEN:
        openingMorphological(image, out, element);
        break;
    case mrcv::METOD_MORF::CLOSE:
        closingMorphological(image, out, element);
        break;
    case mrcv::METOD_MORF::DILAT:
        dilationMorphological(image, out, element);
        break;
    case mrcv::METOD_MORF::ERODE:
        erodeMorphological(image, out, element);
        break;
    case mrcv::METOD_MORF::GRADIENT:
        gradientMorphological(image, out, element);
        break;
    }
return 0;

}
	
	
	
}

File: python/mrcv/mrcv_morphologyimage.py
import cv2
import numpy as np
from enum import Enum
from pathlib import Path

class METOD_MORF(Enum):
    OPEN = 1
    CLOSE = 2
    DILAT = 3
    ERODE = 4
    GRADIENT = 5


def opening_morphological(image: np.ndarray, out: str, element: np.ndarray) -> int:
    """
    @brief Функция морфологического открытия.
    @param image Исходное изображение (в градациях серого).
    @param out Путь для сохранения обработанного изображения.
    @param element Структурирующий элемент для морфологической операции.
    @return Результат работы функции (0 при успехе).
    """
    output = cv2.morphologyEx(image, cv2.MORPH_OPEN, element, iterations=2)
    cv2.imwrite(out, output)
    return 0


def closing_morphological(image: np.ndarray, out: str, element: np.ndarray) -> int:
    """
    @brief Функция морфологического закрытия.
    @param image Исходное изображение (в градациях серого).
    @param out Путь для сохранения обработанного изображения.
    @param element Структурирующий элемент для морфологической операции.
    @return Результат работы функции (0 при успехе).
    """
    output = cv2.morphologyEx(image, cv2.MORPH_CLOSE, element, iterations=2)
    cv2.imwrite(out, output)
    return 0


def gradient_morphological(image: np.ndarray, out: str, element: np.ndarray) -> int:
    """
    @brief Функция морфологического градиента.
    @param image Исходное изображение (в градациях серого).
    @param out Путь для сохранения обработанного изображения.
    @param element Структурирующий элемент для морфологической операции.
    @return Результат работы функции (0 при успехе).
    """
    output = cv2.morphologyEx(image, cv2.MORPH_GRADIENT, element, iterations=1)
    cv2.imwrite(out, output)
    return 0


def erode_morphological(image: np.ndarray, out: str, element: np.ndarray) -> int:
    """
    @brief Функция морфологической эрозии.
    @param image Исходное изображение (в градациях серого).
    @param out Путь для сохранения обработанного изображения.
    @param element Структурирующий элемент для морфологической операции.
    @return Результат работы функции (0 при успехе).
    """
    output = cv2.erode(image, element, iterations=1)
    cv2.imwrite(out, output)
    return 0


def dilation_morphological(image: np.ndarray, out: str, element: np.ndarray) -> int:
    """
    @brief Функция морфологического расширения.
    @param image Исходное изображение (в градациях серого).
    @param out Путь для сохранения обработанного изображения.
    @param element Структурирующий элемент для морфологической операции.
    @return Результат работы функции (0 при успехе).
    """
    output = cv2.dilate(image, element, iterations=1)
    cv2.imwrite(out, output)
    return 0


def morphology_image(image: np.ndarray, out: str, method: METOD_MORF, morph_size: int) -> int:
    """
    @brief Главная функция для применения морфологических операций к изображению.
    @param image Исходное изображение (в градациях серого).
    @param out Путь для сохранения обработанного изображения.
    @param method Метод морфологической операции (OPEN, CLOSE, DILAT, ERODE, GRADIENT).
    @param morph_size Размер структурирующего элемента.
    @return Результат работы функции (0 при успехе).
    """
    element = cv2.getStructuringElement(
        cv2.MORPH_RECT,
        (2 * morph_size + 1, 2 * morph_size + 1),
        (morph_size, morph_size)
    )

    if method == METOD_MORF.OPEN:
        return opening_morphological(image, out, element)
    elif method == METOD_MORF.CLOSE:
        return closing_morphological(image, out, element)
    elif method == METOD_MORF.DILAT:
        return dilation_morphological(image, out, element)
    elif method == METOD_MORF.ERODE:
        return erode_morphological(image, out, element)
    elif method == METOD_MORF.GRADIENT:
        return gradient_morphological(image, out, element)

    return 0